protocol_rfcs/collated-string-type.md

```diff
@@ -8,12 +8,33 @@ This protocol change adds support for collated strings. It consists of three cha
 * Domain metadata with active collation version
 
 --------
+> *** Add New Section after the [Clustered Table](#clustered-table) section***
+# Collations Table Feature
 
-> ***Add a new section in front of the [Primitive Types](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types) section.***
+To support this feature:
+* The table must have Writer Version 7. A feature name `collations` must exist in the table's `writerFeatures`.
+* The feature `domainMetadata` is required in the table protocol's `writerFeatures`. 
+
+## Reader Requirements for Collations:
+
+When Collations is supported (when the `writerFeatures` field of a table's protocol action contains `collations`), then:
+- Readers could do comparisons and sorting of strings based on the collation specified in the schema. The choice of collation and collation version is up to the reader.
+- If the collation is not specified for a string type, then the reader must use the default utf8 binary collation.
 
-### Collations table feature
+## Writer Requirements for Collations:
+
+When Collations is supported (when the `writerFeatures` field of a table's protocol action contains `collations`), then:
+- Writers must write the collation identifier in the schema metadata for a column with non-default collation (utf8 binary).
+- Writer must not write the collation identifier in the schema metadata for a column with default collation (utf8 binary).
+- Writers could write per-file statistics for string columns with collations other than utf8 binary collation in `statsWithCollation`. See [Per-file Statistics](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#per-file-statistics) for more details.
+- Writers could collect per-file statistics for collation versions that are specified in the `domainMetadata` for the `collations` table feature.
+- Writers could remove a collation version from the `domainMetadata` for the `collations` table feature if the collation version is no longer present in the statistics of any AddFile.
```

Comment:
Why there should be no `AddFile` actions with statistics for this collation to allow removal of the collation hint from the `domainMetadata` Doesn't this cause a potential chicken-egg problem with concurrent transactions?

Reply:
On the other hand, if any writer can freely remove the collation version hint from the table, then two different writers can overwrite each other's preferences. Consider this scenario:
* Writer A for engine 1 wants to collect stats for version 120, writer B for engine 2 wants to collect stats for version 121, and writer C for engine 3 can collect stats for version 120 and 121.
* There is no obligation for writer A to collect stats for version 121 nor for writer B to collect stats for version 120 if your writeHints are [120, 121].
* We want writer C to collect stats for both version 120 and 121 because they're both "still in use".
* Writer A should not remove version 121 from the domain metadata and Writer B should not remove version 120, because they are used by other writers.

Among writer A, B, and C, who has more decision power over what collation versions can be dropped from the DomainMetadata? I'd like to avoid a scenario where we are flip-flopping between writeHints for [120] and [121] because writer A and writer B both think the other version is not in use. 
* We could propose tie breaking with preference for newer collation versions, but this may not be reasonable if the user values legacy compatibility. I don't think the Delta protocol should define these rules.
* We could go for semantics where 1) any writer can freely remove collation versions from the domain metadata but 2) any writer _must_ update the collation versions in the domain metadata if it adds an AddFile that has statsWithCollations with a specific version. In this case, the domain metadata serves almost as a LRU cache.
* We could have this be inconsistent (i.e. the version hints can have versions not present in the AddFile stats and the AddFiles can have stats that are not present in the version hints) and not require anything else, but then what is the purpose of the version hints in this case? What guarantees do the version hints provides? What work does it prevent the writers from doing? In this scenario, I would argue that we should just remove the hints from the format. The decision of what collation versions to collect stats for is completely up to the underlying engine.

Reply:
I think the scenario that has to be legal is: There is version A, then we add B and collect stats for A and B, after backfilling stats for B we remove version A (also from AddFiles to save space).

Now what we want to forbid is difficult indeed. One thing I can see is that writers must not write stats for a version that is not in the DomainMetadata.